LightPositionTool : Add tool for placing shadows

ericmehl opened the pull request:
This adds a new tool to the scene viewer for placing shadows. Two points are set using `Control` + Click to set the pivot point and `Shift` + Click to set the shadow point. The selected light (or any selection, like a group) is repositioned so it lies along the line from the shadow point to the pivot, the same distance back from the pivot as its original distance from the pivot. It's also reoriented to face the shadow point.

This currently uses the last selected object as the source of the original position. I'll add a follow up commit to use the centroid of the full selection.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

ericmehl commented:
When I use the shadow placement tool and then try and Undo, it doesn't work if the undo is targeted to the viewport. But it does if the mouse is hovering over the node graph. I think I'm getting the `ScriptNode` in the standard way, but maybe not. Currently it doesn't behave as the other transform tools do.

ericmehl commented:
I fixed the undo behavior and in the process simplified the modifier logic. I had been tracking the "setting pivot" / "setting shadow point" state via `keyPressSignal()`, but handling `Control` was preventing `Control + Z` from getting its signal.

I found that `MouseEvent` has it's own keyboard modifier state, so I'm using that now.

ericmehl commented:
I don't quite have transformations right in the case of multiple lights. My latest attempt is in https://github.com/GafferHQ/gaffer/pull/5558/commits/48bc79fb68d0fedd367ee8e0abfe504272c6743f, but I don't quite have it yet. I think it's either `centroidTransform` on line 114 not being calculated right, or it's a problem with `offset` on line 133.

johnhaddon commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -0,0 +1,347 @@
+		)
+	)
+	{
+		return false;
```
When testing, I got into a few unfortunate situations where I ended up positioning something that wasn't a light. I think part of the problem was this `return false`, which would let me accidentally select some geometry with a left click. Once I'd done that, my immediate instinct was to `Ctrl + Click` to deselect the bit of geometry, but of course that's setting a pivot, which moves the selected object.

I'm not sure what the solution is, but I wonder if we need to either expose _all_ the usual selection behaviours, or _none_ of them? If the former, then we'd need different keypresses, in which case perhaps `V` could work for the "shadow point" click, somewhat matching the snapping behaviours in the TranslateTool and RotateTool?

johnhaddon replied:
I'm not sure if it's related to this or not, but I've also been seeing a bunch of the following :

```WARNING : TransformTool::updateSelection : Last selected path not included in selection```

Here's a simple repro :

1. Open Gaffer.
2. Make a Cube node and view its output.
3. Enable the LightPositionTool.
4. Left-click the cube in the Viewer.

ericmehl replied:
Got to the bottom of this one - it's a result of what I was doing in my `selectionChangedSignal()` handler :
1. `SelectionTool` handles a click event and calls `ContextAlgo::setSelectedPaths()`, _before_ `ContextAlgo::setLastSelectedPath()` https://github.com/GafferHQ/gaffer/blob/2d22be25ee4aabe54f85cb9a8dce4e7bf1f4dd7f/src/GafferSceneUI/SelectionTool.cpp#L241
2. Setting the selected paths fires my `selectionChangedSignal()` handler
3. I was calling `TransformTool::selection()`, which does `updateSelection()`
4. `updateSelection()` assumes we have a proper `lastSelectedPath` and gives the warning noted above if there's nothing there https://github.com/GafferHQ/gaffer/blob/2d22be25ee4aabe54f85cb9a8dce4e7bf1f4dd7f/src/GafferSceneUI/TransformTool.cpp#L1180-L1192

I think this will all be solved when I add an actual handle, so I'll move my centroid code to `updateHandles()` where I think it is maybe more appropriate.

I don't know if there's anything to be improved with the `selectionChangedSignal()` behavior - it seems logical that in that handler I could ask for the current selection? But that's not part of this effort.

johnhaddon replied:
Good sleuthing!

> I don't know if there's anything to be improved with the selectionChangedSignal() behavior - it seems logical that in that handler I could ask for the current selection?

Yep, completely. The current behaviour is breaking one of the fundamental rules of signals - don't emit them while your visible state is inconsistent.

> I'll move my centroid code to updateHandles() where I think it is maybe more appropriate.

Yep, that makes sense (although I think we're dropping the centroid anyway aren't we?). I suspect the reason this hasn't become a problem in the other tools is because they are following the "don't do any work unless visible" rule, so everything is deferred until `preRender()`, at which point the selection state is consistent.

johnhaddon commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -0,0 +1,347 @@
+
+	for( const auto &s : selection() )
+	{
+		// See `RotateTool::buttonPress()` for a description of why we use this relatively
+		// elaborate orientation calculation.
```
I'm not sure the orientation stuff is working great in this context. I tested with a spotlight with a gobo, and although the initial click keeps the gobo upright, after a while it has drifted off so it's rotated on its side. You can see the same thing with a quadlight - use a second viewer to look through the quadlight while positioning it in the other viewer, and you'll see the look-through get a twist on it.

I'm not sure what the behaviour should be if the light isn't upright (local Y aligned-ish with world Y) in the first place, but if it _is_ upright it feels to me that it should remain that way (i.e. local Y is as aligned to world Y as it can be while still achieving the required aim).

johnhaddon replied:
I'm still finding this to be the case in the latest version.

johnhaddon commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -0,0 +1,347 @@
+	m_shadowPoint( std::nullopt )
+{
+	SceneGadget *sg = runTimeCast<SceneGadget>( this->view()->viewportGadget()->getPrimaryChild() );
+	// We have to insert this before the underlying SelectionTool connections or it starts an object drag.
+	sg->buttonPressSignal().connectFront( boost::bind( &LightPositionTool::buttonPress, this, ::_2 ) );
```
It'd be nice to support drags for the shadow placement - having placed the pivot already, it would be great to get continuous feedback from a drag rather than making a series of clicks. If we did that, it seems like we might as well support the same for the pivot as well?

johnhaddon commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -0,0 +1,347 @@
+		return true;
+	}
+
+	return input == scenePlug()->transformPlug();
```
Do we need this? The documentation says we only need to consider things we use in `updateHandles()`.

johnhaddon commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -0,0 +1,347 @@
+			return;
+		}
+
+		m_worldCentroid = V3f( 0 ) * selection().back().orientedTransform( Orientation::World );
```
I've been getting a few of these errors in my terminal :

```ERROR : Emitting signal : Selection is not editable```

Not 100% sure it's coming from here, but `orientedTransform()` is one of the functions that could throw that exception (in TransformTool.cpp).

johnhaddon commented on resources/graphics.svg:
The general idea behind the icon seems pretty reasonable (more applicable to the upcoming specular-position mode, but still workable for the general tool). But the lines seem very thick compared to the other icons - perhaps could do with a bit of refinement?

johnhaddon commented on python/GafferSceneUI/LightPositionToolUI.py:
```diff
@@ -0,0 +1,72 @@
+
+	"nodeToolbar:bottom:type", "GafferUI.StandardNodeToolbar.bottom",
+
+	"toolbarLayout:customWidget:InteractionTipWidget:widgetType", "GafferSceneUI.LightPositionToolUI._InteractionTipWidget",
```
The extra height on this is affecting the vertical positioning of the "Editing ..." widget on the left, so it doesn't match the other transform tools. Not sure what the solution would be - I wonder if the tooltip could even go in the "Editing..." widget? We seems to have a mix of overlay-based (LightTool) tooltips and Widget-based (here) tooltips, and I wonder if there's any merit in trying to unify things a bit. Or if maybe the tools are different enough that it makes sense...

johnhaddon reviewed (commented):
Thanks Eric, this is going to be very nice! 

I haven't reviewed the maths side of this much at all, because you indicated you want to rework the centroid stuff, and my comments on the up-vector stuff might change things too. @danieldresser-ie might do a better job of reviewing anyway, once we've settled on the behaviour we want.

I mentioned in one comment that it'd be good to have some visual indicators for the pivot and shadow point. Here's a very quick straw-man mockup to get the conversation startup on that :

![image](https://github.com/GafferHQ/gaffer/assets/1133871/d345254b-6c3a-467d-9496-625ad1304218)

I suggest we spend a bit of time in today's meeting getting feedback from the rest of the team, and settling on something a bit more concrete...

Cheers...
John

ericmehl commented:
My latest push has some new functionality :
- The tip of the arrow (cone) is repositioned so the apex of the cone is at the shadow target. Before the base of the cone was placed there which, besides being not quite correct, I think contributed to the handle going into hover state when you placed the target.
- I moved the viewport tip explaining how to place the target and pivot to the top.
- The other transform tools have a similar tip about using the `V` key.
- You can now drag the line from light to shadow target to change the distance of the light along that line. Anywhere you drag on the line will move the light relative to the start of the drag point. Is this discoverable enough? Would it be better to add a translate handle onto the light origin?
- Target and pivot points are stored based on the selection. They are now only cleared if you select a light that is not along the shadow line and pointing at the shadow target.
- Undo now applies to placing the target and pivot points as well as the actual light transformation. This is handy because of the way I'm resetting the points. Without this, if you have a shadow line, undo the last transform, then you'll also reset the shadow line since the light isn't in the place the shadow handle expects it to be.

A couple more items to go :
- Hold `V` and drag to continually move the shadow target. Probably for the pivot too?
- Rotate handle on the end of the shadow handle to rotate the light around the Z axis.
- Perhaps better orientation when placing the light?

ericmehl commented on python/GafferSceneUI/TransformToolUI.py:
```diff
@@ -101,6 +105,22 @@ def _distance( ancestor, descendant ) :
+				with GafferUI.ListContainer( orientation = GafferUI.ListContainer.Orientation.Vertical, spacing = 4 ) :
+					for label in labels:
+						GafferUI.Label( label )
+		else :
```
I haven't been able to sort out why this is not hiding the tooltip widget in the case of activating the scale tool, which doesn't do targeting. Is there a subwidget other than `self` that I should be hiding instead? I always seem to get a small box regardless of what I've tried hiding.

murraystevenson commented:
Awesome! This is really coming along. Here's a bit of a laundry list of nitpicky usability thoughts after trying out the latest:

Top centre could be a bit distracting for the instruction placement. They're the sort of thing you'd want to look past once you "get it" and that's a little hard to do front and centre. Maybe off to the left is better so they're more associated with the tools themselves? I'm also seeing the rest of the Viewer's top toolbar jump a bit when tools are deactivated and the instructions are no longer visible.

The translate/rotate tool instruction language of "place translation/orientation target" feels a bit off, "place target" suggests a persistent target rather than a one-off action. Maybe language along the lines of "Hold V and click to position on target" "Hold V and click to aim at target" is more clear?

We can currently translate a light past the shadow pivot while dragging along the visualiser line - I'm not sure if that makes sense to do so? Adjusting the pivot or target point afterwards results in the light moving back to the equivalent distance on the "far" side of the pivot. The visualisation line also breaks once the light passes beyond the shadow target point. 

It'd be nice to have distinct cursor icon for pivot placement as a sanity check before you go and place the wrong thing. Undo definitely helps here, but I still find myself going through some mental gymnastics to ensure I'm doing the right thing while switching between pivot and target placement. A distinct icon could reinforce the learning of the behaviour and allow me to correct "in flight" rather than make a mistake, undo it, and try again...

Maybe we should make the pivot visualisation a sphere rather than a ring for less visual clutter, the line running through the ring reads a bit like a :no_entry_sign:.

The line also runs through the end of the target cone on hover. Maybe this contributes to the visualisation getting in the way of making fine adjustments to the target placement? 

![image](https://github.com/GafferHQ/gaffer/assets/50844517/20789a13-eb70-4e4b-9930-9f2b5638ce2b)

> You can now drag the line from light to shadow target to change the distance of the light along that line. Anywhere you drag on the line will move the light relative to the start of the drag point. Is this discoverable enough? Would it be better to add a translate handle onto the light origin?

I kind of want to reserve judgment on this until I see the "drag to continually move the pivot/target" behaviour. Though I think whatever we do here should be considered in terms of consistency with any other light placement tool...

ericmehl commented:
My latest push wraps up the final features for the shadow part of the light position tool. I added the rotate handle and continuous placement during drag, as well as a few fixes.

I will go through the code once more time to to see if there are any loose ends before ticking the Ready for review button.

ericmehl commented:
I tidied up the code a bit and pushed that up. Throwing the switch to open it for review!

ericmehl commented:
Pushed a small fixup to address a bad optional access bug when using the rotate handle without first dragging the shadow / pivot distance handle.

murraystevenson commented:
Dragging the shadow target around a raytraced viewport is a lot of fun! The other `V` placement tools are missing out...

One thing noticed in usability testing is that dragging a target generates two undo events, one for the initial click and then a second for the drag, so you end up having to undo twice to revert to the previous shadow placement. Maybe not a huge deal, but it caught me off-guard the first time I tried undo-ing a dragged shadow target placement. The (single) undo didn't put the light back where it had started, as my click to instigate the drag was some distance away from the original shadow target position.

johnhaddon commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -0,0 +1,969 @@
+		m_shadowTargetMap.erase( m_key );
+		m_shadowPivotMap.erase( m_key );
+		m_shadowPivotDistanceMap.erase( m_key );
+	}
```
It doesn't seem possible to use the LightPositionTool when there's a Group transforming the light downstream. In the script below you can see this by focussing the Group and then trying to use the tool on `/group/spotLight`. I'm not 100% sure this chunk of code is the culprit, but I suspect so - it seems very much like the pivot/target are getting rejected as soon as they've been set.

This sort of "edit above a downstream transform" thing was very hard to get right for the other tool types - I'd suggest adding some test coverage to give us a safety net.

<details>

```
import Gaffer
import GafferArnold
import GafferScene
import imath

Gaffer.Metadata.registerValue( parent, "serialiser:milestoneVersion", 1, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:majorVersion", 3, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:minorVersion", 8, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:patchVersion", 0, persistent=False )

__children = {}

__children["Cube"] = GafferScene.Cube( "Cube" )
parent.addChild( __children["Cube"] )
__children["Cube"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Plane"] = GafferScene.Plane( "Plane" )
parent.addChild( __children["Plane"] )
__children["Plane"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Parent"] = GafferScene.Parent( "Parent" )
parent.addChild( __children["Parent"] )
__children["Parent"]["children"].addChild( GafferScene.ScenePlug( "child1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Parent"]["children"].addChild( GafferScene.ScenePlug( "child2", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Parent"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["SpotLight"] = GafferArnold.ArnoldLight( "SpotLight" )
parent.addChild( __children["SpotLight"] )
__children["SpotLight"].loadShader( "spot_light" )
__children["SpotLight"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"] = GafferScene.Group( "Group" )
parent.addChild( __children["Group"] )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Cube"]["transform"]["scale"].setValue( imath.V3f( 1, 5.59656191, 1 ) )
__children["Cube"]["__uiPosition"].setValue( imath.V2f( -5.08920288, 3.61759186 ) )
__children["Plane"]["transform"]["rotate"].setValue( imath.V3f( -90, 0, 0 ) )
__children["Plane"]["transform"]["scale"].setValue( imath.V3f( 155.261353, 155.261353, 155.261353 ) )
__children["Plane"]["__uiPosition"].setValue( imath.V2f( 7.91079521, 3.61778164 ) )
__children["Parent"]["in"].setInput( __children["Cube"]["out"] )
__children["Parent"]["parent"].setValue( '/' )
__children["Parent"]["children"][0].setInput( __children["Plane"]["out"] )
__children["Parent"]["children"][1].setInput( __children["SpotLight"]["out"] )
__children["Parent"]["__uiPosition"].setValue( imath.V2f( 8.41027451, -4.546875 ) )
__children["SpotLight"]["name"].setValue( 'spotLight' )
__children["SpotLight"]["transform"]["translate"].setValue( imath.V3f( -6.37608194, 3.91808772, 0.26422596 ) )
__children["SpotLight"]["transform"]["rotate"].setValue( imath.V3f( -114.534935, -95.2462158, 103.170578 ) )
__children["SpotLight"]["parameters"]["exposure"].setValue( 7.0 )
__children["SpotLight"]["__uiPosition"].setValue( imath.V2f( 20.9107952, 3.61795092 ) )
__children["Group"]["in"][0].setInput( __children["Parent"]["out"] )
__children["Group"]["transform"]["rotate"].setValue( imath.V3f( 0, 90, 0 ) )
__children["Group"]["__uiPosition"].setValue( imath.V2f( 9.9087038, -12.7127008 ) )


del __children
```


</details>

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5558/commits/7a3cf3b4f419b81929e0ac68856f01e4c3bc8594 and https://github.com/GafferHQ/gaffer/pull/5558/commits/2df67e588fdefb72a6188c5a94f7b1d7bb5b3759

ericmehl commented:
> One thing noticed in usability testing is that dragging a target generates two undo events, one for the initial click and then a second for the drag, so you end up having to undo twice to revert to the previous shadow placement.

Good catch - that is odd behavior. I fixed it in https://github.com/GafferHQ/gaffer/pull/5558/commits/75236a802750b98165b859fc48db2cd315443631. Perhaps using `TransformTool::dragEnd()` is a bit questionable, since I'm doing it in `buttonRelease()`, but it's an expeditious way to increment `m_mergeGroupId` which is private to `TransformTool` and button presses and drags aren't completely unrelated, I figure.

johnhaddon commented on python/GafferSceneUI/TranslateToolUI.py:
```diff
@@ -51,6 +53,8 @@
 	"viewer:shortCut", "W",
 	"order", 1,
 
+	"ui:tool:transformToolTips", IECore.StringVectorData( [ "Hold 'V' and click to move to target", ] ),
```
I think "snap to geometry" would have a more familiar ring than "move to target" for most people.

ericmehl replied:
Changed in https://github.com/GafferHQ/gaffer/pull/5558/commits/ed4eb1588f632803363529cdc831140ffed18f4c

johnhaddon commented on python/GafferSceneUI/TransformToolUI.py:
```diff
@@ -101,6 +109,23 @@ def _distance( ancestor, descendant ) :
+
+		GafferUI.Frame.__init__( self, borderWidth = 4, **kw )
+
+		labels = Gaffer.Metadata.value( tool, "ui:tool:transformToolTips" ) or []
```
We have `tool:` for stuff that applies to all tools - I think `transformTool:toolTips` might sit better with that.

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5558/commits/174cef119594ecc1ccec5d8ad7c0c9ddd4b1f063

johnhaddon commented on python/GafferSceneUI/TransformToolUI.py:
```diff
@@ -101,6 +109,23 @@ def _distance( ancestor, descendant ) :
+		labels = Gaffer.Metadata.value( tool, "ui:tool:transformToolTips" ) or []
+
+		with self :
+			with GafferUI.Frame( borderWidth = 0 ) as self.__innerFrame :
+				with GafferUI.ListContainer( orientation = GafferUI.ListContainer.Orientation.Vertical, spacing = 4 ) :
```
A Frame, a ListContainer and N Labels seems a bit overkill here. Can we just use a single Label with `"\n".join( labels )`?

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5558/commits/3aef5f2522e17d0c72a0026f2f6a9eebb45bd5c8

johnhaddon reviewed (changes requested):
Thanks Eric, as Murray says, this is getting to be fun! Comments inline as usual...

johnhaddon commented:
I've also started monkeying around with our Qt stuff to see if I can track down the cause of the annoying jiggle-on-deactivate. Haven't got anywhere yet (but I think did find the cause of another Qt annoyance) but will keep digging...

ericmehl commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -0,0 +1,969 @@
+	{
+		return;
+	}
+	const Selection s = selection().back();
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5558/commits/f2f3f9779e3b27a96d1acf6f307ae9fb1ae40864

ericmehl commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -0,0 +1,969 @@
+	auto shadowHandle = static_cast<ShadowHandle *>( m_shadowHandle.get() );
+	Action::enact(
+		scriptNode,
+		[this, p, shadowHandle]() {
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5558/commits/cad54844a33b63eb36238339a0e852da9243df17

ericmehl commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -0,0 +1,969 @@
+	Action::enact(
+		scriptNode,
+		[this, p, shadowHandle]() {
+			m_shadowTargetMap[m_key] = p;
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5558/commits/cad54844a33b63eb36238339a0e852da9243df17

ericmehl commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -0,0 +1,969 @@
+
+	auto shadowHandle = static_cast<ShadowHandle *>( m_shadowHandle.get() );
+
+	m_key = MurmurHash().append( s.path() );
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5558/commits/cad54844a33b63eb36238339a0e852da9243df17

ericmehl commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -0,0 +1,969 @@
+			dirty( DirtyType::Render );
+		}
+
+		std::optional<V3f> getShadowTarget() const
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5558/commits/ddd6aeaa806f066e7bbb1a4b0b8e651d1744976d

ericmehl commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -0,0 +1,969 @@
+			);
+
+			auto standardStyle = runTimeCast<const StandardStyle>( style );
+			assert( standardStyle );
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5558/commits/5ca9d03303d5255da2f1ecc928a3cac39862fecf

johnhaddon replied:
This isn't really a fix - it just means we'll crash on the next line rather than have a failed assertion.

ericmehl replied:
Head smack, let's try https://github.com/GafferHQ/gaffer/pull/5558/commits/2df3050bec8abb05a1cb12f018d97658c98be25f instead. I default to the what is currently the hard-coded color in `StandardStyle` if `style` is not `StandardStyle`.

ericmehl commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -0,0 +1,969 @@
+				new IECoreGL::Color(
+					enabled() ? ( highlighted ? g_lightToolHighlightColor4 : highlightColor4 ) : g_lightToolDisabledColor4
+				)
+			);
```
I reversed the colors in https://github.com/GafferHQ/gaffer/pull/5558/commits/518ce36a690ce11044c8ac846dea781c987ca46a. I think it is better in this, reversed, way. The handle matches the light and makes that visual connection, and the blue hover state makes the drag-ability more obvious.

My recollection of the LightTool coloring is that making the handles yellow made them hard to see the change when the tool is activated, making it harder to realize you could drag the handles. Especially for the spot light angle tools, where the dot icons are fairly subtle.

The quad light handles are maybe easier to understand since they have arrows / cones at the edges.

But I think that decision also came at a time when we didn't have the tooltip that follows the mouse for the light tool. So maybe after this is sorted it's worth revisiting to see how a color reversal there looks.

ericmehl commented:
Thanks John! I addressed the feedback for the easy stuff today. I think there are a number of variables and methods that can be renamed to remove the shadowy-ness from them and prep for the highlight placement tool. It will just have a target, no pivot, so I don't see any particular reason why I can't drop `shadowTarget` wording for `target` and similar throughout.

Also next up is looking at the transforms which I'll tuck into in the morning.

ericmehl commented on include/GafferSceneUI/LightPositionTool.h:
```diff
@@ -0,0 +1,137 @@
+		bool affectsHandles( const Gaffer::Plug *input ) const override;
+		void updateHandles( float rasterScale ) override;
+
+	private :
```
I think just a `placementMode()` plug, which is a method rather than member data. Will that count as an ABI break? I could add it now and have it be a no-op until the specular mode is in place.

Otherwise I think the specular mode will have everything it needs - a target, the normals come from the buffer, viewer position comes from the viewer, and I can reuse what I'm now calling `DistanceHandle` (the handle formerly known as `ShadowHandle`).

johnhaddon replied:
New plugs aren't an ABI break, so sounds like all is well.

ericmehl commented on include/GafferSceneUI/LightPositionTool.h:
```diff
@@ -0,0 +1,137 @@
+
+		GAFFER_NODE_DECLARE_TYPE( GafferSceneUI::LightPositionTool, LightPositionToolTypeId, TransformTool );
+
+		void position( const Imath::V3f &shadowPivot, const Imath::V3f &shadowTarget, const float targetDistance );
```
I renamed `position()` to `positionShadow()` in https://github.com/GafferHQ/gaffer/pull/5558/commits/9917937dc24e08ead979f26350cf010ab371bbab and renamed a few other things in that commit. I think we can generally use `pivot` and `target` nomenclature and reuse the same concepts for shadow and highlight placement.

ericmehl commented:
I pushed a few new changes that I believe take care of all the remaining comments and improvements. I also updated the commit references my response to the comments - rebasing caused the commit hashes to change.

The fix to "undo" when drag placing the target / pivot is now in https://github.com/GafferHQ/gaffer/pull/5558/commits/2540da48216083d1b27553914b8c0000c68ad887.

Ready for a new look!

ericmehl commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -0,0 +1,969 @@
+bool LightPositionTool::keyPress( const KeyEvent &event )
+{
+	// We track this regardless of whether we're active or not in case the tool
+	// is changed whilst the key is held down.
```
I copied this from `RotateTool` without really thinking through the accuracy, but I believe in both cases the comment is incorrect and should be removed - tracking is in fact conditioned on active state. Or is my post-holiday brain not fully online and I'm mistaken now?

I don't think removing the condition is right since that would cause the cursor to change. Probably just need to remove the comment I think.

johnhaddon replied:
Removing the comment looks good to me.

ericmehl replied:
Done in https://github.com/GafferHQ/gaffer/pull/5558/commits/d1d01f938346dd42dc12cbd3567ef70124a8784b

johnhaddon commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -969,22 +973,26 @@ void LightPositionTool::setShadowTarget( const V3f &p, ScriptNodePtr scriptNode
 std::optional<V3f> LightPositionTool::getShadowTarget() const
 {
-	auto it = m_shadowTargetMap.find( m_key );
+	auto it = m_shadowTargetMap.find( ScenePlug::pathToString( selection().back().path() ) );
```
Again, `selection().back()` is being called here without checking that the selection is non-empty, and we're doing the same thing in a bunch of other places in this file. It's _possible_ that the check is being performed by the callers in all cases, but I haven't verified - in any case it seems fragile. I wonder if we should just use a little `selectedPathString()` utility function that returns `""` when there's no selection?

ericmehl replied:
I added the helper function in https://github.com/GafferHQ/gaffer/pull/5558/commits/53faf38062e7d976b24f39beb424bc63f7541f8b. I also double checked that the remaining `selection().back()` calls are protected behind a check for multiple items in `selection()`. All are, via `TransformTool::selectionEditable()` or checking that `selection().size() == 1`.

johnhaddon commented on python/GafferSceneUI/LightPositionToolUI.py:
```diff
@@ -0,0 +1,60 @@
+	"order", 7,
+	"tool:exclusive", True,
+
+	"ui:transformTool:toolTips", IECore.StringVectorData( [ "Hold 'V' to place shadow target", "Hold 'Shift' + 'V' to place shadow pivot", ] ),
```
I think it makes sense for the pivot tooltip to go first, since the natural order of operations is to set that first. I'm also wondering if there's a good reason this is StringVectorData (that we join with "\n" later) rather than just being a single string (with `\n` in it)?

ericmehl replied:
Right, that was a remnant from when I was doing multiple label widgets. Fixed in https://github.com/GafferHQ/gaffer/pull/5558/commits/6b1ed74ae79ab186f708defeb550512793713967

johnhaddon reviewed (commented):
Thanks for the update Eric. Got a couple of final comments inline, including one crash that we definitely will need to address. Nearly there though!

ericmehl commented:
One thing I didn't about regarding the preventing an ABI break when I add highlight placement mode - I'm planning on adding an enum to indicate the tool mode. That would be public so we can bind it in Python.

Would that consittute an ABI break? I'm a little fuzzy on what counts as an ABI break, is it _just_ adding new data members?

ericmehl commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -969,22 +973,26 @@ void LightPositionTool::setShadowTarget( const V3f &p, ScriptNodePtr scriptNode
-			shadowHandle->setShadowTarget( currentValue );
+		[t = LightPositionToolPtr( this ), k = pathString, currentValue, shadowHandle]() {
+			t->m_shadowTargetMap[k] = currentValue;
+			if( ScenePlug::pathToString( t->selection().back().path() ) == k )
```
Same as above

ericmehl commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -942,22 +942,26 @@ void LightPositionTool::setShadowPivot( const V3f &p, ScriptNodePtr scriptNode )
-			shadowHandle->setShadowPivot( currentValue );
+		[t = LightPositionToolPtr( this ), k = pathString, currentValue, shadowHandle]() {
+			t->m_shadowPivotMap[k] = currentValue;
+			if( ScenePlug::pathToString( t->selection().back().path() ) == k )
```
This is fixed now using the helper function introduced https://github.com/GafferHQ/gaffer/pull/5558/commits/53faf38062e7d976b24f39beb424bc63f7541f8b to make sure there is at least one item in `selection()`.

ericmehl commented:
I have all the comments from today addressed - I think 
* reviewing those commits
* double checking we're happy with the behavior
* considering the `enum` ABI break question
* getting Arnold downloads working again for CI to check out

are what's left.

johnhaddon commented:
Thanks Eric, LGTM!

> I'm a little fuzzy on what counts as an ABI break, is it just adding new data members?

There's also changing existing function signatures, removing data members, changing virtual functions, and a bunch of other subtlety I don't have the best handle on. Basically, stuff that changes the layout of a class in memory, or the calling convention of a function. Adding a new enum type doesn't affect any of that, so should be fine.

> reviewing those commits

Done!

> double checking we're happy with the behavior

I haven't done as extensive testing as @murraystevenson here - did you want one last check Murray or are you happy?

> getting Arnold downloads working again for CI to check out

I hope I should have this taken care of by the time you get in 🤞.

One other small thing - I needed this patch to build on Mac :

```
--- a/include/GafferSceneUI/LightPositionTool.h
+++ b/include/GafferSceneUI/LightPositionTool.h
@@ -45,6 +45,8 @@
 
 #include "Gaffer/ScriptNode.h"
 
+#include <unordered_map>
+
 namespace GafferSceneUI
 {
 
diff --git a/src/GafferSceneUI/LightPositionTool.cpp b/src/GafferSceneUI/LightPositionTool.cpp
index 27af71ea6..ede376de1 100644
--- a/src/GafferSceneUI/LightPositionTool.cpp
+++ b/src/GafferSceneUI/LightPositionTool.cpp
@@ -685,8 +685,6 @@ void LightPositionTool::updateHandles( float rasterScale )
        V3f direction;
        transform.multDirMatrix( V3f( 0, 0, -1.f ), direction );
 
-       const V3f handleDir = ( shadowTarget.value() - shadowPivot.value() ).normalized();
-
        if(
                !m_drag &&
                (
```

Perhaps you could incorporate that while squashing down ready for merging?

ericmehl commented:
Squashed and double checked that nothing was changed in the process. And included the fix for mac, thanks for pointing that out.

Once we're happy with the user-facing details, should be ready for a merge!

murraystevenson commented on python/GafferSceneUI/TransformToolUI.py:
```diff
@@ -101,6 +109,21 @@ def _distance( ancestor, descendant ) :
+
+		GafferUI.Frame.__init__( self, borderWidth = 4, **kw )
+
+		label = Gaffer.Metadata.value( tool, "ui:transformTool:toolTips" ) or ""
```
Maybe this should be "ui:transformTool:toolTip" seeing as it's now a single string?

ericmehl replied:
Yes, that should be singular now, fixed and squashed into https://github.com/GafferHQ/gaffer/pull/5558/commits/f22a800cf4f8c291440786a844ec7f399322180e

murraystevenson commented on resources/graphics.py:
```diff
@@ -33,6 +33,7 @@
 				"pointerAdd",
 				"pointerRemove",
 				"pointerRotate",
+                "pointerPivot",
```
We've ended up with spaces here.

ericmehl replied:
That's no good! Fixed and squashed into https://github.com/GafferHQ/gaffer/pull/5558/commits/18b122489faf389ee7ab5dcad65544b4d7b70021

murraystevenson reviewed (commented):
Thanks Eric! This looks good to me for a release. I've added a couple of very pedantic comments inline, but otherwise seems good to merge. :+1:

ericmehl commented:
Thanks for the double check @murraystevenson! I fixed those two issues and squashed it down to their relevant commits. Should be ready for a merge now.

johnhaddon commented:
Thanks Eric!
